tpu_inference/distributed/tpu_connector_local.py

```diff
@@ -858,43 +987,62 @@ def _save_blocks_to_cpu(self, req_id: ReqId, full_block_ids: list[int],
                          "KV caches not registered.")
             return req_id
 
-        num_tokens_to_save = len(
-            full_token_ids) - save_spec.skip_leading_tokens
+        blocks_to_save = save_spec.src_blocks
+        num_total_tokens = save_spec.num_total_tokens
+        num_skip_leading_tokens = save_spec.num_skip_leading_tokens
+
+        assert num_total_tokens <= len(
+            full_token_ids), f"{num_total_tokens} > {len(full_token_ids)}"
+
+        num_tokens_to_save = num_total_tokens - num_skip_leading_tokens
         if num_tokens_to_save <= 0 and not save_spec.is_final_save:
             logger.info(f"Request {req_id}: No new tokens to save.")
             return req_id
 
-        tokens_to_process = full_token_ids[save_spec.skip_leading_tokens:]
-
-        # Calculate the block slice.
-        start_block_idx = save_spec.skip_leading_tokens // self.block_size
-        blocks_to_process = full_block_ids[start_block_idx:]
+        process_token_ids = full_token_ids[:num_total_tokens]
```

Comment:
we can combine the two steps to one 
tokens_to_save = full_token_ids[num_skip_leading_tokens: num_total_tokens]